# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [165]:
# Imports section
import pandas as pd 
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import numpy as np

## 1. Loading the dataset

In [166]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
file_path = 'science_data_large.csv' 
data = pd.read_csv(file_path)

print(data.head(15))

print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [167]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X = data[['Mols KCL', 'Temperature °C']]  # Selecting features
y = data['Size nm^3']  # Assuming Temperature is the label

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [168]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()
model.fit(X_train, y_train)

sample_datapoint = pd.DataFrame({'Mols KCL': [600], 'Temperature °C': [400]})  # Use representative values
predicted_output = model.predict(sample_datapoint)

# Report the model's score
model_score = model.score(X_test, y_test)

# Extract coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

# Printing the results (if needed)
print(f"Predicted output: {predicted_output}")
print(f"Model score: {model_score}")
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted output: [556684.12565684]
Model score: 0.8552472077276096
Coefficients: [1032.69506649  866.14641337]
Intercept: -409391.4795834087


Write the linear equation of a slime: (example equation: $E = mc^2$)
$h(x) = -409391.4795834087 + 1032.69506649(Mols KCL) + (866.14641337)(Temperature C)$

Report on score and explain meaning: The score is about 74%, which indicates the model can explain about 85.5% of the variance in Temperature C by the linear regression model. This suggests it is a good fit for the data. 

## 4. Use Cross Validation

In [169]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance
scores = cross_val_score(model, X, y, cv=5, scoring='r2', n_jobs=-1)

# Report the findings
print(f"Cross-validated R² scores: {scores}")
print(f"Mean R² score: {scores.mean()}")
print(f"Standard deviation of R² scores: {scores.std()}")

# Calculate the mean and standard deviation of the scores
mean_score = scores.mean()
std_dev_score = scores.std()

Cross-validated R² scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean R² score: 0.8568167899144437
Standard deviation of R² scores: 0.01346630737209602


Write findings here: This means that on average, the model can explain about 85.68% variance in Temperature C. This means there is a good relationship between Mols KCL, Temperature C and Size nm^3. The low standard deviation also means the performance is consistent across different folds. 

## 5. Using Polynomial Regression

In [170]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.
X = data[['Mols KCL', 'Temperature °C']]
y = data['Size nm^3']

# Create polynomial features of degree 2
poly = PolynomialFeatures(degree=2, include_bias=False)  # do not include bias (constant term)
X_poly = poly.fit_transform(X)

# Initialize the polynomial regression model
model_poly = LinearRegression()

# Create a KFold object with shuffling
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform k-fold cross-validation with shuffling
scores_poly = cross_val_score(model_poly, X_poly, y, cv=kf, scoring='r2', n_jobs=-1)

# Report the findings
print(f"Cross-validated R² scores for polynomial regression: {scores_poly}")
print(f"Mean R² score for polynomial regression: {scores_poly.mean()}")
print(f"Standard deviation of R² scores for polynomial regression: {scores_poly.std()}")

# Fit the model on the entire dataset
model_poly.fit(X_poly, y)

# Extract coefficients and intercept
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

# Output resultant equation
# Generate feature names for the polynomial features
feature_names = poly.get_feature_names_out(input_features=['Mols KCL', 'Temperature °C'])
equation_terms = [f"{coeff} * {name}" for coeff, name in zip(coefficients_poly, feature_names)]
equation = " + ".join(equation_terms)

print(f"Resultant polynomial equation:\nSize nm^3 = {intercept_poly} + {equation}")

Cross-validated R² scores for polynomial regression: [1. 1. 1. 1. 1.]
Mean R² score for polynomial regression: 1.0
Standard deviation of R² scores for polynomial regression: 0.0
Resultant polynomial equation:
Size nm^3 = 1.65727105922997e-05 + -1.2311232548171453e-07 * Mols KCL + 12.000000002870134 * Temperature °C + 0.02857142870391094 * Mols KCL^2 + 2.0000000000135154 * Mols KCL Temperature °C + -1.0566436614567465e-11 * Temperature °C^2


Write the polynomial equation of a slime: (example equation: $E = mc^2$) $Size nm^3 = 1.65727105922997e-05 + -1.2311232548171453e-07 * Mols KCL + 12.000000002870134 * Temperature °C + 0.02857142870391094 * Mols KCL^2 + 2.0000000000135154 * Mols KCL Temperature °C + -1.0566436614567465e-11 * Temperature °C^2$

Report on the score and interpret: The score is about 1, which is a perfect fit, meaning that it suggests better predictive ability. This means the model can perfectly predict the size based on the temperature Cand mols KCL. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [171]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

# Load the CKD dataset
file_path = 'ckd_feature_subset.csv'  # Ensure this matches the filename you have
data = pd.read_csv(file_path)

# Display the first few rows of the dataset and info for inspection
print(data.head())
print(data.info())

# Separating features and target variable
# Adjust 'Target_ckd' to the actual name in your dataset if necessary
X = data.drop('Target_ckd', axis=1)  # Replace 'Target_ckd' with the appropriate target column name
y = data['Target_ckd']

# Check for any missing values in the features
print(X.isnull().sum())  # Check for null values

# Define the classification models
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Support Vector Machine': SVC(),
    'k-Nearest Neighbors': KNeighborsClassifier(),
}

# Initialize a list to store results
results = []

# Random seed for reproducibility
random_state = 42

# Create a KFold object with shuffling
kf = KFold(n_splits=5, shuffle=True, random_state=random_state)

# Train and evaluate the models
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=kf, scoring='accuracy')  # Use 'accuracy' as scoring
    results.append({
        'Model': name,
        'Mean Accuracy': scores.mean(),
        'Standard Deviation': scores.std()
    })

# Define neural network configurations to test
nn_configs = [
    ('NN_1', MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=random_state)),
    ('NN_2', MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=random_state)),
    ('NN_3', MLPClassifier(hidden_layer_sizes=(50, 50), max_iter=1000, random_state=random_state))  # 2 hidden layers
]

# Evaluate each neural network configuration
for name, nn_model in nn_configs:
    nn_scores = cross_val_score(nn_model, X, y, cv=kf, scoring='accuracy')
    results.append({
        'Model': name,
        'Mean Accuracy': nn_scores.mean(),
        'Standard Deviation': nn_scores.std()
    })

# Compile results into a DataFrame
results_df = pd.DataFrame(results)

# Display the results
print(results_df)

# Optional: Exporting to a CSV for reference
results_df.to_csv('ckd_classification_results.csv', index=False)

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB
None
age 

c:\Users\coold\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\coold\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\coold\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\coold\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the 

                    Model  Mean Accuracy  Standard Deviation
0     Logistic Regression       0.856559            0.066269
1  Support Vector Machine       0.928172            0.047601
2     k-Nearest Neighbors       0.927957            0.052440
3                    NN_1       0.908602            0.052041
4                    NN_2       0.935054            0.040466
5                    NN_3       0.954624            0.043696


## Results and Conclusion for Classification Experiments
The classification experiments indicated that the best performing model was the neural network configuration NN_3, achieving a mean accuracy of 0.9546 with a standard deviation of 0.0437. This model, featuring two hidden layers with 50 neurons each, effectively captured complex patterns in the CKD dataset, showcasing its robust predictive capability for classifying Chronic Kidney Disease. Other models, such as the Support Vector Machine (mean accuracy of 0.9282) and k-Nearest Neighbors (mean accuracy of 0.9280), also demonstrated strong performance, indicating their suitability for this task. In contrast, the Logistic Regression model (mean accuracy of 0.8566) performed significantly worse, highlighting its limitations in capturing the non-linear relationships present in the data.

## Results and Conclusion for Neural Network Experiments
The analysis of the neural network configurations revealed that the choice of architecture significantly impacted classification performance. Specifically, the number of hidden layers and neurons per layer were critical factors; NN_3 outperformed the simpler NN_1 (with a single hidden layer of 10 neurons) and NN_2 (with a single layer of 100 neurons). Additionally, the activation function and the number of training iterations  are essential parameters that can affect model training and convergence. These findings suggest that utilizing a deeper architecture and optimizing training configurations can substantially enhance the model’s ability to classify chronic kidney disease effectively.